eventually-v0-5-0/src/version.rs

```diff
@@ -0,0 +1,52 @@
+use std::{
+    fmt::{Debug, Display, Formatter, Result as FmtResult},
+    ops::Add,
+};
+
+use serde::{Deserialize, Serialize};
+
+#[derive(Debug, Default, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
+pub struct Version(pub u64);
+
+impl Display for Version {
+    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
+        Debug::fmt(self, f)
+    }
+}
+
+impl Add<u64> for Version {
+    type Output = Self;
+
+    fn add(self, other: u64) -> Self::Output {
+        Self(self.0 + other)
+    }
+}
+
+impl Add for Version {
+    type Output = Self;
+
+    fn add(self, other: Self) -> Self::Output {
+        self + other.0
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, thiserror::Error)]
+#[error("conflict error detected, expected event stream version was: {expected}, found: {actual}")]
+pub struct ConflictError {
+    pub expected: Version,
+    pub actual: Version,
+}
+
+pub trait ToConflictError {
+    fn to_conflict_error(&self) -> Option<ConflictError> {
+        None
+    }
+}
+
+impl ToConflictError for ConflictError {
+    fn to_conflict_error(&self) -> Option<ConflictError> {
+        Some(*self)
+    }
+}
+
+impl ToConflictError for std::convert::Infallible {}
```

Comment:
it looks like you're reinventing `TryFrom`

Reply:
This is a good point! Rather than `TryFrom`, this is more of a `From<Option<ConflictError>>`, right?

Reply:
it seems to be that

```rust
t.to_conflict_error()
```
is exactly equivalent to

```rust
t.try_into::<ConflictError>().ok()
```

with the benefit that you're using a standard API and all the ecosystem benefits that go with it

this example looks slightly clumsy because of the explicit type annotations, which would not in general be necessary

Reply:
`t.try_into::<ConflictError>().ok()` assumes an error in the parsing, which is not the case. It's a downcast, which either returns `Some(ConflictError)` or `None`, depending on whether the downcasting succeeded.

Thanks for pointingit out, I'm changing the bound to `Into<Option<ConflictError>>` 💯